src/background/tab-state.js

```diff
@@ -3,48 +3,61 @@ import { RequestCanceledError } from './errors';
 
 import * as uriInfo from './uri-info';
 
-const states = {
-  ACTIVE: 'active',
-  INACTIVE: 'inactive',
-  ERRORED: 'errored',
-};
+/**
+ * @typedef {import('./direct-link-query').Query} Query
+ */
+
+/**
+ * Hypothesis-related state for a specific tab.
+ *
+ * This should really be named `TabState` but that will conflict with the
+ * `TabState` class below. That class needs to be renamed first.
+ *
+ * @typedef State
+ * @prop {'active'|'inactive'|'errored'} state - Whether or not H is active on the page
+ * @prop {number} annotationCount -
+ *   The count of annotations on the page visible to the user, as returned by
+ *   the badge API
+ * @prop {boolean} extensionSidebarInstalled - Has the client been loaded into this tab?
+ * @prop {boolean} ready - Is the tab loaded and ready for the client to be loaded?
+ * @prop {Error} [error]
+ * @prop {Query} [directLinkQuery]
+ */
 
-/** The default H state for a new browser tab */
+/**
+ * The default H state for a new browser tab.
+ *
+ * @type {State}
+ */
 const DEFAULT_STATE = {
-  // Whether or not H is active on the page
-  state: states.INACTIVE,
-  // The count of annotations on the page visible to the user,
-  // as returned by the badge API
+  state: 'inactive',
   annotationCount: 0,
-  // Whether or not the H sidebar has been installed onto the page by
-  // the extension
   extensionSidebarInstalled: false,
-  // Whether the tab is loaded and ready for the sidebar to be installed.
   ready: false,
-  // The error for the current tab.
   error: undefined,
 };
 
-/** TabState stores the H state for a tab. This state includes:
- *
- * - Whether the extension has been activated on a tab
- * - Whether the sidebar is currently installed on a tab
- * - The count of annotations visible to the user on the URL currently
- *   displayed in the tab.
- *
- * The H state for a tab is updated via the setState() method and
- * retrieved via getState().
- *
- * When the H state for a tab changes, the `onchange()` callback will
- * be triggered with the tab ID and current and previous states.
+/**
+ * @typedef {{ [tabId: number]: State }} TabStateMap
+ */
+
+/**
+ * TabState stores the Hypothesis-related state for tabs in the current browser
+ * session.
  *
- * initialState - An Object of tabId/state keys. Used when loading state
- *   from a persisted store such as localStorage. This will be merged with
- *   the default state for a tab.
- * onchange     - A function that receives onchange(tabId, current).
+ * @param {TabStateMap} initialState - Initial state information for tabs, eg.
+ *   from a persistent store.
+ * @param {(tabId: number, current: undefined|State) => any} onchange -
+ *   Callback invoked when state for a tab changes
  */
 export default function TabState(initialState, onchange) {
```

Comment:
Line 77 could we remove if we add this default value.

```suggestion
export default function TabState(initialState, onchange = null) {
```

Reply:
Indeed. I've gone one step further and just marked the parameter as optional. There was no particular reason to initialize the `onchange` field of this class to `null` rather than `undefined` if not set from what I can see.